jt_notes.md

```diff
@@ -0,0 +1,127 @@
+
+app
+
+For my benefit: base_camera.py runs a background thread which posts events whenever a frame is available. frames() uses "yield" to pass frame buffers to an iterator.
+
+app.gen() is a "video streaming generator function" which in turn "yield"s frames (in what looks like a MIME-like format) to the caller. I am not clear how video_feed behaves, though.
+
+Basic questions **for chas**:
+
+- What is the overall function of app.py? Do I understand correctly that these decorated functions are called in response to messages from the client webpage? In response to those messages, they appear to both perform internal state-affecting actions (i.e. run the synchronization algorithms, in some cases), *and also* then return the actual HTTP source(*) that should now be displayed in the webpage. Am I understanding that correctly? What is the deal with the "session" object? Could really do with a comment at the top of the module to explain what's going on with all of this.                (*) Oh, or sometimes raw video images apparently, in video_feed()...
+
+- **IMPORTANT:** When app.run() refers to "live", is that what actually gets run for live sync with a microscope, fish and everything? Because, for instance, it calls through to **emulate**_get_period...!? I don't know if this is because, as you said, the app code needs updating, but I can't see how it is meaningfully interacting with cli in the case of **live** frames, and I can't quite picture how that even *should* work. It seems like the gen() function should be being used somehow, but that's not being used except in video_feed().
+
+- base_camera.py needs a similar comment at the top of the module explaining what's going on - and in particular what the relationship is between a "client" and a "thread", explanation of the role of "ident" (which is a unique identifier associated with each thread), etc. Are these concurrent threads or [I suspect] sequential threads? Really needs an explanation of the thread architecture here, and the role of "yield" in terms how of the threads interact.
+
+- **IMPORTANT:** the same questions apply to the Flask app. As per my comment in app.py, I am starting to suspect that this could be running concurrent threads. That needs careful thought and documenting, because it gets very complicated.
+
+- More generally, *as far as I can see*, there doesn't seem to be to be a critical need to be using [sequential] threads here at all (though it doesn't do any harm). Out of interest, is this code just adapted from some online example? I suspect it makes a lot more sense in the case where there is *more than one* client involved (but I don't think that's really an issue for us in reality).
```

Comment:
What bit are talking about here? The only bit that was based on another codebase was the streaming page, everything else was purely us I think?

Reply:
Not a big deal, but I think I was referring to a suspicion that there is not actually anything being achieved by BaseCamera._thread() that couldn't be achieved by polling for new frames from a single thread. I could very well be wrong there, but if so then I have not wrapped my head around what that benefit is, exactly.